### PR TITLE
Added EVP_PKEY_check and EVP_PKEY_public_check with support for EC and RSA keys

### PR DESCRIPTION
**Pull request by smittals2**
### Description of changes: 
Added EVP_PKEY_check and EVP_PKEY_public_check with support for EC and RSA keys and tests. Added check and public check function definitions in EVP_PKEY_METHOD and EVP_PKEY_ASN1_METHOD. Defined RSA (rsa_pkey_check & rsa_pkey_public_check) and EC (ec_pkey_check and ec_pkey_public_check) specific wrappers. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

**Comment by dkostic** [comment]
Is it possible to avoid expanding `pmeth` and `ameth` structures and instead just call the `RSA/EC_check_key` functions directly from the top level `EVP_PKEY_check` and `EVP_PKEY_public_check`?

File: crypto/evp_extra/evp_asn1.c [hunk 1]
```diff
@@ -274,6 +274,48 @@ static EVP_PKEY *old_priv_decode(CBS *cbs, int type) {
   return NULL;
 }
 
+int EVP_PKEY_check(EVP_PKEY_CTX *ctx) {
+  EVP_PKEY *pkey = ctx->pkey;
```
Comment (dkostic) [inline crypto/evp_extra/evp_asn1.c:278]: check `ctx != NULL`?

File: crypto/evp_extra/evp_asn1.c [hunk 2]
```diff
@@ -274,6 +274,48 @@ static EVP_PKEY *old_priv_decode(CBS *cbs, int type) {
...
+
+  if (pkey == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NO_KEY_SET);
+    return 0;
+  }
+
+  switch(pkey->type) {
+    case EVP_PKEY_EC:
+      if (EC_KEY_get0_private_key(pkey->pkey.ec) == NULL) {
```
Comment (dkostic) [inline crypto/evp_extra/evp_asn1.c:287]: why do we need this?

File: crypto/evp_extra/evp_asn1.c [hunk 3]
```diff
@@ -274,6 +274,48 @@ static EVP_PKEY *old_priv_decode(CBS *cbs, int type) {
...
+        OPENSSL_PUT_ERROR(EVP, EC_R_MISSING_PRIVATE_KEY);
+        return 0;
+      }
+      return EC_KEY_check_key(pkey->pkey.ec);
+    case EVP_PKEY_RSA:
+      return RSA_check_key(pkey->pkey.rsa);
+    default:
+      OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
+      return -2;
```
Comment (dkostic) [inline crypto/evp_extra/evp_asn1.c:296]: why -2, we usually return 0 for failures?

File: crypto/ecdh_extra/ecdh_test.cc [hunk 4]
```diff
@@ -104,6 +104,17 @@ TEST(ECDHTest, TestVectors) {
     ASSERT_TRUE(EC_KEY_set_public_key(key.get(), pub_key.get()));
     ASSERT_TRUE(EC_KEY_check_key(key.get()));
 
+    // Check EVP_PKEY_check and EVP_PKEY_public_check
```
Comment (andrewhop) [inline crypto/ecdh_extra/ecdh_test.cc:107]: Do we have any tests with bad keys we can add this too?

File: include/openssl/evp.h [hunk 5]
```diff
@@ -698,6 +698,16 @@ OPENSSL_EXPORT int EVP_PKEY_derive_set_peer(EVP_PKEY_CTX *ctx, EVP_PKEY *peer);
 OPENSSL_EXPORT int EVP_PKEY_derive(EVP_PKEY_CTX *ctx, uint8_t *key,
                                    size_t *out_key_len);
 
+// |EVP_PKEY_check| validates the key-pair given by |ctx|. If the key type is supported,
```
Comment (andrewhop) [inline include/openssl/evp.h:701]: Style required for docs:
```suggestion
// EVP_PKEY_check validates the key-pair given by |ctx|. If the key type is supported,
```

File: crypto/evp_extra/evp_asn1.c (hunk 2 above)
Comment (smittals2) [reply crypto/evp_extra/evp_asn1.c:292]: EVP_PKEY_check is supposed to validate both the private and public aspects of the key as opposed to EVP_PKEY_public_check which only validates the public component of the key pair. EC_KEY_check_key [doesn't fail ](https://github.com/aws/aws-lc/blob/13337f78937bc5911905e71d58ce9c71c23166ee/crypto/fipsmodule/ec/ec_key.c#L310)when priv_key is null, so there is an additional check before.

File: crypto/ecdh_extra/ecdh_test.cc (hunk 4 above)
Comment (dkostic) [reply crypto/ecdh_extra/ecdh_test.cc:107]: we can add `EVP_PKEY_check` to every test where `RSA_check_key` and `EC_KEY_check_key` are  tested. that would cover "bad" keys.

File: crypto/evp_extra/evp_asn1.c (hunk 2 above)
Comment (dkostic) [reply crypto/evp_extra/evp_asn1.c:292]: RSA_check_key also doesn't fail when given a public key only.

Comment (dkostic) [reply crypto/evp_extra/evp_asn1.c:292]: > EVP_PKEY_check is supposed to validate both the private and public aspects

so OpenSSL fails when `EVP_PKEY_check` is called with public key only?

File: crypto/fipsmodule/ec/ec_test.cc [hunk 6]
```diff
@@ -1023,6 +1024,16 @@ TEST(ECTest, ArbitraryCurve) {
   // The key must be valid according to the new group too.
   EXPECT_TRUE(EC_KEY_check_key(key2.get()));
 
+  EVP_PKEY *ec_pkey = EVP_PKEY_new();
```
Comment (dkostic) [inline crypto/fipsmodule/ec/ec_test.cc:1027]: we usually use `bssl::UniquePtr` for pointers in the C++ code. did you try that?

File: include/openssl/evp.h [hunk 7]
```diff
@@ -697,6 +697,18 @@ OPENSSL_EXPORT int EVP_PKEY_derive_set_peer(EVP_PKEY_CTX *ctx, EVP_PKEY *peer);
 OPENSSL_EXPORT int EVP_PKEY_derive(EVP_PKEY_CTX *ctx, uint8_t *key,
                                    size_t *out_key_len);
 
+// EVP_PKEY_check supports EC and RSA keys and wraps the corresponding key check functions.
+// In OpenSSL, this function is meant to validate the key-pair, however, our key checking
+// logic is less restrictive in that it allows keys with only a public component. To avoid introducing
+// inconsistencies in key checking behavior, this function is implemented differently than in OpenSSL.
+// |EVP_PKEY_check| validates the public component of the key and private component if available.
+// Returns one on success and 0 on error.
```
Comment (dkostic) [inline include/openssl/evp.h:705]:
```suggestion
// Returns one on success and zero on error.
```

File: include/openssl/evp.h [hunk 8]
```diff
@@ -697,6 +697,18 @@ OPENSSL_EXPORT int EVP_PKEY_derive_set_peer(EVP_PKEY_CTX *ctx, EVP_PKEY *peer);
...
+OPENSSL_EXPORT int EVP_PKEY_check(EVP_PKEY_CTX *ctx);
+
+// EVP_PKEY_public_check wraps |EVP_PKEY_check|. Validates the public component of the key and private component if
+// provided. Returns one on success and 0 on error.
```
Comment (dkostic) [inline include/openssl/evp.h:709]:
```suggestion
// provided. Returns one on success and zero on error.
```

File: include/openssl/evp.h [hunk 9]
```diff
@@ -697,6 +697,18 @@ OPENSSL_EXPORT int EVP_PKEY_derive_set_peer(EVP_PKEY_CTX *ctx, EVP_PKEY *peer);
 OPENSSL_EXPORT int EVP_PKEY_derive(EVP_PKEY_CTX *ctx, uint8_t *key,
                                    size_t *out_key_len);
 
+// EVP_PKEY_check supports EC and RSA keys and wraps the corresponding key check functions.
```
Comment (dkostic) [inline include/openssl/evp.h:700]: nit: we try to limit line length to 80 characters.

File: crypto/fipsmodule/ec/ec_test.cc (hunk 6 above)
Comment (andrewhop) [reply crypto/fipsmodule/ec/ec_test.cc:1027]: The sanitizers and valgrind flagged these as leaked.

**Review by dkostic — APPROVED** [review]

File: crypto/rsa_extra/rsa_test.cc [hunk 10]
```diff
@@ -839,52 +867,80 @@ TEST(RSATest, CheckKey) {
...
+
   // Missing n or e does not pass.
   ASSERT_TRUE(BN_hex2bn(&rsa->n, kN));
   EXPECT_FALSE(RSA_check_key(rsa.get()));
+
+  bssl::UniquePtr<EVP_PKEY_CTX> rsa_key_ctx(
+          EVP_PKEY_CTX_new(rsa_pkey.get(), NULL));
+  ASSERT_TRUE(rsa_key_ctx);
+  EXPECT_FALSE(EVP_PKEY_check(rsa_key_ctx.get()));
+  EXPECT_FALSE(EVP_PKEY_public_check((rsa_key_ctx.get())));
```
Comment (andrewhop) [inline crypto/rsa_extra/rsa_test.cc:882]: Is this test separate than what's on 875? Should the clear error be moved up there?

File: crypto/evp_extra/evp_asn1.c [hunk 11]
```diff
@@ -274,6 +274,34 @@ static EVP_PKEY *old_priv_decode(CBS *cbs, int type) {
...
+    return 0;
+  }
+
+  switch(pkey->type) {
+    case EVP_PKEY_EC:
+      return EC_KEY_check_key(pkey->pkey.ec);
+    case EVP_PKEY_RSA:
+      return RSA_check_key(pkey->pkey.rsa);
+    default:
+      OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
```
Comment (andrewhop) [inline crypto/evp_extra/evp_asn1.c:296]: NP: do you have a test for this case? Such as a kyber key.

**Review by andrewhop — APPROVED** [review]

**Review by dkostic — APPROVED** [review]

**Pull request closed without merging**